### PR TITLE
Climb code

### PR DESCRIPTION
**Pull request by Aadit-Ambadkar**
For manav.

### CODE REVIEW

**Comment by noraatfedora** [comment]
@Maanav-s make sure you know how to submit pull requests on your own

**Review by noraatfedora — COMMENTED** [review]
In the future use the standard git commit naming convention
https://namingconvention.org/git/commit-message-naming.html

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -34,15 +42,21 @@ public RobotContainer() {
...
 
-  /**
-   * Use this to pass the autonomous command to the main {@link Robot} class.
-   *
-   * @return the command to run in autonomous
-   */
+  private void configureButtonBindings() {
+    new JoystickButton(opController, Constants.climbUp)
+      .whenPressed(new RunCommand(climb::raiseClimbStart, climb))
+      .whenReleased(climb::stopClimb, climb);
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: whenReleased methods also require commands, not just lambdas

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import com.ctre.phoenix.motorcontrol.TalonFXInvertType;
+
+
+
+public class Climb extends SubsystemBase {
+
+    WPI_TalonFX left = new WPI_TalonFX(Constants.climbLeftCAN);
+    WPI_TalonFX right = new WPI_TalonFX(Constants.climbRightCAN);
+
+    Solenoid brakeSolenoid = new Solenoid(PneumaticsModuleType.CTREPCM, Constants.climbSolenoidChannel);
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: we use the DoubleSolenoid class for pneumatics (see 2021 intake for how to do it)

File: takoyaki/src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -4,12 +4,21 @@
...
+
+public final class Constants {
+    public static int
+        climbUp = XboxController.Button.kY.value,
+        climbDown = XboxController.Button.kA.value,
+        climbBrakeOn = XboxController.Button.kX.value,
+        climbBrakeOff = XboxController.Button.kB.value,
+        
+        climbSolenoidChannel = 1,
+        climbLeftCAN = 15,
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/Constants.java]: we generally name the constants LEFT_MOTOR_ID

File: takoyaki/src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -4,12 +4,21 @@
...
- *
- * <p>It is advised to statically import this class (or one of its inner classes) wherever the
- * constants are needed, to reduce verbosity.
- */
-public final class Constants {}
+import edu.wpi.first.wpilibj.XboxController;
+
+public final class Constants {
+    public static int
+        climbUp = XboxController.Button.kY.value,
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/Constants.java]: all constants are should be public static final and all caps with an underscore between words

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -7,24 +7,32 @@
...
+ 
 public class RobotContainer {
-  // The robot's subsystems and commands are defined here...
+
   private final ExampleSubsystem m_exampleSubsystem = new ExampleSubsystem();
 
   private final ExampleCommand m_autoCommand = new ExampleCommand(m_exampleSubsystem);
 
-  /** The container for the robot. Contains subsystems, OI devices, and commands. */
+  XboxController opController = new XboxController(0);
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: operator controller is always in port 1
the port should also be a constant in the constants file

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -7,24 +7,32 @@
...
+  Climb climb = new Climb();
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: all subsystem objects in robot container are private final

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    boolean brake = false;
+
+  public Climb() {
+    left.configFactoryDefault();
+    right.configFactoryDefault();
+
+    left.follow(right);
+
+    left.setInverted(TalonFXInvertType.OpposeMaster);
+    right.setInverted(TalonFXInvertType.Clockwise);
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: set the leader motor inversion before the follower

File: takoyaki/src/main/java/frc/robot/Constants.java [hunk 8]
```diff
@@ -4,12 +4,21 @@
...
+        climbBrakeOn = XboxController.Button.kX.value,
+        climbBrakeOff = XboxController.Button.kB.value,
+        
+        climbSolenoidChannel = 1,
+        climbLeftCAN = 15,
+        climbRightCAN = 15
+    ;
+    public static double
+        climbMaxUp = 0.5,
+        climbMaxDown = 1
```
Comment (poroheart) [inline takoyaki/src/main/java/frc/robot/Constants.java]: one of these values should be negative to run the motor in the other direction

File: takoyaki/src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (KTOmega) [reply takoyaki/src/main/java/frc/robot/Constants.java]: yes, please do not instantiate multiple at a time, please do one at a time.

e.g.

```
public static final int CLIMB_UP = XboxController.Button.kY.value;
public static final int CLIMB_DOWN = XboxController.Button.kA.value;
```

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 9]
```diff
@@ -7,24 +7,32 @@
...
 import frc.robot.commands.ExampleCommand;
+import frc.robot.subsystems.Climb;
 import frc.robot.subsystems.ExampleSubsystem;
 import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.RunCommand;
+import edu.wpi.first.wpilibj2.command.button.JoystickButton;
 
 /**
+ *
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: remove unnecessary addition

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -7,24 +7,32 @@
...
 
 /**
+ * 
  * This class is where the bulk of the robot should be declared. Since Command-based is a
  * "declarative" paradigm, very little robot logic should actually be handled in the {@link Robot}
  * periodic methods (other than the scheduler calls). Instead, the structure of the robot (including
  * subsystems, commands, and button mappings) should be declared here.
  */
+
+
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: nit: no extra empty lines before class decl

File: takoyaki/src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (KTOmega) [reply takoyaki/src/main/java/frc/robot/RobotContainer.java]: yes, no magic numbers. also, please specify visibility

```suggestion
  private final XboxController opController = new XboxController(Constants.kOI.DRIVE_CONTROLLER);
```

File: takoyaki/src/main/java/frc/robot/RobotContainer.java [hunk 11]
```diff
@@ -34,15 +42,21 @@ public RobotContainer() {
    * edu.wpi.first.wpilibj.Joystick} or {@link XboxController}), and then passing it to a {@link
    * edu.wpi.first.wpilibj2.command.button.JoystickButton}.
    */
-  private void configureButtonBindings() {}
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/RobotContainer.java]: nit: no extra lines before method decl

File: takoyaki/src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (KTOmega) [reply takoyaki/src/main/java/frc/robot/RobotContainer.java]:
```suggestion
      .whenReleased(new RunCommand(climb::stopClimb, climb));
```

this should've produced a build error, was this built before submission?

Comment (KTOmega) [reply takoyaki/src/main/java/frc/robot/RobotContainer.java]: same on line 52

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.TalonFXInvertType;
+
+
+
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: nit: remove unnecessary whitespace

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
+import edu.wpi.first.wpilibj.Solenoid;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: nit: remove unnecessary whitespace

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
+import edu.wpi.first.wpilibj.Solenoid;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: nit: remove unnecessary whitespace

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.TalonFXInvertType;
+
+
+
+public class Climb extends SubsystemBase {
+
+    WPI_TalonFX left = new WPI_TalonFX(Constants.climbLeftCAN);
+    WPI_TalonFX right = new WPI_TalonFX(Constants.climbRightCAN);
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: please specify visibility operators, i.e. `private final`

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+  @Override
+  public void periodic() {
+    
+  }
+
+  @Override
+  public void simulationPeriodic() {
+  }
+
+  public void raiseClimbStart(){
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: nit: one space before opening brace
```suggestion
  public void raiseClimbStart() {
```

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    right.set(ControlMode.PercentOutput, Constants.climbMaxDown);
+  }
+  public void stopClimb(){
+    right.set(ControlMode.PercentOutput, 0);
+  }
+  public void triggerBrake(){
+    brakeSolenoid.set(true);
+  }
+  public void releaseBrake(){
+      brakeSolenoid.set(false);
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: nit: incorrect tab spacing

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+
+public class Climb extends SubsystemBase {
+
+    WPI_TalonFX left = new WPI_TalonFX(Constants.climbLeftCAN);
+    WPI_TalonFX right = new WPI_TalonFX(Constants.climbRightCAN);
+
+    Solenoid brakeSolenoid = new Solenoid(PneumaticsModuleType.CTREPCM, Constants.climbSolenoidChannel);
+
+    boolean brake = false;
```
Comment (KTOmega) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: unused member variable

**Review by KTOmega — DISMISSED** [review]

File: takoyaki/src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (nguyen-b) [reply takoyaki/src/main/java/frc/robot/Constants.java]: Even better if you can name them something like `CLIMB_UP_BUTTON`

File: takoyaki/src/main/java/frc/robot/Constants.java (hunk 3 above)
Comment (nguyen-b) [reply takoyaki/src/main/java/frc/robot/Constants.java]: Or `LEFT_DRIVE_MOTOR_CAN_ID`. Tells left motor on what subsystem and what type of ID.

File: takoyaki/src/main/java/frc/robot/subsystems/Climb.java (hunk 2 above)
Comment (nguyen-b) [inline takoyaki/src/main/java/frc/robot/subsystems/Climb.java]: Team Preference is to assign these in constructor.

**Comment by Aadit-Ambadkar** [comment]
@Maanav-s also re-request reviews until you get the pr accepted.
